### PR TITLE
Add mTLS support for upstream servers

### PR DESCRIPTION
**Pull request by periklis**
The following PR adds support to reverse proxy calls to metrics/logs/traces endpoints using mTLS. The purpose for this to support for example a Loki cluster that is secured via mTLS on an isolated environment. In detail, Loki does not support any native authentication methods on the HTTP layer.

### CODE REVIEW

File: api/logs/v1/http.go [hunk 1]
```diff
@@ -131,6 +131,10 @@ func NewHandler(read, tail, write *url.URL, upstreamCA []byte, opts ...HandlerOp
 				t.TLSClientConfig.RootCAs.AppendCertsFromPEM(upstreamCA)
 			}
 
+			if upstreamCert != nil {
+				t.TLSClientConfig.Certificates = append(t.TLSClientConfig.Certificates, *upstreamCert)
```
Comment (xperimental) [inline api/logs/v1/http.go:135]: If `upstreamCA` is empty then `TLSClientConfig` is still `nil` here. In that case it will panic here.

File: api/logs/v1/http.go [hunk 2]
```diff
@@ -229,6 +233,10 @@ func NewHandler(read, tail, write *url.URL, upstreamCA []byte, opts ...HandlerOp
...
```
Comment (xperimental) [inline api/logs/v1/http.go:237]: `TLSClientConfig` can be `nil`.

File: api/logs/v1/http.go [hunk 3]
```diff
@@ -275,6 +283,10 @@ func NewHandler(read, tail, write *url.URL, upstreamCA []byte, opts ...HandlerOp
...
```
Comment (xperimental) [inline api/logs/v1/http.go:287]: `TLSClientConfig` can be `nil`.

File: api/metrics/legacy/http.go [hunk 4]
```diff
@@ -120,9 +120,14 @@ func NewHandler(url *url.URL, upstreamCA []byte, opts ...HandlerOption) http.Han
 			t.TLSClientConfig = &tls.Config{
 				RootCAs: x509.NewCertPool(),
 			}
+
 			t.TLSClientConfig.RootCAs.AppendCertsFromPEM(upstreamCA)
 		}
 
+		if upstreamCert != nil {
+			t.TLSClientConfig.Certificates = append(t.TLSClientConfig.Certificates, *upstreamCert)
```
Comment (xperimental) [inline api/metrics/legacy/http.go:128]: `TLSClientConfig` can be `nil`.

File: api/metrics/v1/http.go [hunk 5]
```diff
@@ -205,6 +205,10 @@ func NewHandler(read, write, rulesEndpoint *url.URL, upstreamCA []byte, opts ...
 				t.TLSClientConfig.RootCAs.AppendCertsFromPEM(upstreamCA)
 			}
 
+			if upstreamCert != nil {
+				t.TLSClientConfig.Certificates = append(t.TLSClientConfig.Certificates, *upstreamCert)
```
Comment (xperimental) [inline api/metrics/v1/http.go:209]: `TLSClientConfig` can be `nil`.

File: api/metrics/v1/http.go [hunk 6]
```diff
@@ -250,6 +254,10 @@ func NewHandler(read, write, rulesEndpoint *url.URL, upstreamCA []byte, opts ...
...
```
Comment (xperimental) [inline api/metrics/v1/http.go:258]: `TLSClientConfig` can be `nil`. I also noticed that this seems to be the only handler where `t` is initialized using `http.DefaultTransport` instead of creating a new one, not part of this PR though.

**Review by xperimental — CHANGES_REQUESTED** [review]
Just one inline comment (repeated though).

Would it make sense to extract the `TLSClientConfig` initialization (and maybe also the CLI-parameter parsing) to a separate function that can be re-used across the handlers?

File: api/traces/v1/http.go [hunk 7]
```diff
@@ -137,6 +139,17 @@ func NewV2Handler(read *url.URL, readTemplate string, opts ...HandlerOption) htt
...
 
+		if len(upstreamCA) != 0 {
+			t.TLSClientConfig = &tls.Config{
+				RootCAs: x509.NewCertPool(),
+			}
+			t.TLSClientConfig.RootCAs.AppendCertsFromPEM(upstreamCA)
+		}
+
+		if upstreamCert != nil {
+			t.TLSClientConfig.Certificates = append(t.TLSClientConfig.Certificates, *upstreamCert)
```
Comment (xperimental) [inline api/traces/v1/http.go:150]: `TLSClientConfig` can be `nil`.

**Comment by periklis** [comment]
> Just one inline comment (repeated though).
> 
> Would it make sense to extract the `TLSClientConfig` initialization (and maybe also the CLI-parameter parsing) to a separate function that can be re-used across the handlers?

Good catch according to our CLI flags we enable upstream TLS only if we provide an upstream CA, thus I wrapped the cert inside the first if block.
